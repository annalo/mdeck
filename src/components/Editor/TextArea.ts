import styled from "styled-components";

const TEXT_AREA_LINE_HEIGHT = 18;

const TextArea = styled.textarea`
  background-color: ${(props) => props.theme.editorBackgroundColor};
  border: none;
  color: ${(props) => props.theme.textColor};
  font-size: 15px;
  height: 100%;
  line-height: ${TEXT_AREA_LINE_HEIGHT}px;
  outline: none;
  padding: 0.5em;
  padding-bottom: 90vh;
  resize: none;
  tab-size: 2;

  ::-webkit-scrollbar {
    display: none;
  }
  -ms-overflow-style: none;
  scrollbar-width: none;
`;

export { TEXT_AREA_LINE_HEIGHT, TextArea };
