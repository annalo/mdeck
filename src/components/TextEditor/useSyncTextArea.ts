import { useEffect } from "react";
import type { RefObject } from "react";
import * as R from "ramda";

interface UseSyncTextAreaProps {
  ref: RefObject<HTMLTextAreaElement>;
  slideshowLineNumber: number;
  textAreaLineHeight: number;
}

export const useSyncTextArea = ({
  ref,
  slideshowLineNumber,
  textAreaLineHeight,
}: UseSyncTextAreaProps): void => {
  /* Syncs text when slideshowLineNumber changes */
  useEffect(() => {
    const getNode = R.prop("current");
    const scrollTop = R.multiply(textAreaLineHeight, slideshowLineNumber);
    // TODO smooth scrolling
    const setScrollTop = R.curry((value, n) => {
      n.scrollTop = value; // eslint-disable-line no-param-reassign
    });

    R.pipe(getNode, R.unless(R.isNil, setScrollTop(scrollTop)))(ref);
  }, [ref, slideshowLineNumber, textAreaLineHeight]);
};
