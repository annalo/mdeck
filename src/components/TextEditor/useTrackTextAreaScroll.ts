import { useEffect, useMemo } from "react";
import * as R from "ramda";
import throttle from "lodash/throttle";

import type { Dispatch, RefObject } from "react";
import {
  MarkdownContextReducerAction,
  MarkdownContextReducerActionType,
} from "types/markdown-context";

interface UseTrackTextAreaScrollProps {
  dispatch: Dispatch<MarkdownContextReducerAction>;
  isActive: boolean;
  ref: RefObject<HTMLTextAreaElement>;
  textAreaLineHeight: number;
}

function useTrackTextAreaScroll({
  dispatch,
  isActive,
  ref,
  textAreaLineHeight,
}: UseTrackTextAreaScrollProps): void {
  const handleScroll = useMemo(
    () =>
      throttle((e) => {
        const getScrollTop = R.path(["target", "scrollTop"]);
        const calculateLineNumber = R.divide(R.__, textAreaLineHeight);
        const textLineNumber = R.pipe(
          getScrollTop,
          calculateLineNumber,
          Math.floor
        )(e);

        dispatch({
          type: MarkdownContextReducerActionType.SetTextLineNumber,
          textLineNumber,
        });
      }, 200),
    [dispatch, textAreaLineHeight]
  );

  /* Adds/Removes event listener on 'scroll' depending on pane `isActive` */
  useEffect(() => {
    const node = ref.current;

    isActive
      ? node?.addEventListener("scroll", handleScroll, { passive: true })
      : node?.removeEventListener("scroll", handleScroll);

    return () => node?.removeEventListener("scroll", handleScroll);
  }, [isActive, handleScroll, ref]);
}

export { useTrackTextAreaScroll };
