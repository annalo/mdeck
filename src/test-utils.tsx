/* eslint-disable import/no-extraneous-dependencies */

import React from "react";
import { render } from "@testing-library/react";
import { renderHook } from "@testing-library/react-hooks";

import { ThemeProvider } from "components/Home/ThemeProvider";
import { MarkdownContextProvider } from "contexts/MarkdownContext";
import { SlideObserverProvider } from "contexts/SlideObserver";
import { CodeLineObserverProvider } from "contexts/CodeLineObserver";

const AllTheProviders = ({ children }) => {
  return (
    <ThemeProvider>
      <MarkdownContextProvider>
        <SlideObserverProvider>
          <CodeLineObserverProvider>{children}</CodeLineObserverProvider>
        </SlideObserverProvider>
      </MarkdownContextProvider>
    </ThemeProvider>
  );
};

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

const customRenderHook = (hook, options) =>
  renderHook(hook, { wrapper: AllTheProviders, ...options });

// override render and renderHook methods
export { customRender as render, customRenderHook as renderHook };
