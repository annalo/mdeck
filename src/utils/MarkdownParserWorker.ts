import Worker from "worker-loader!./Worker"; // eslint-disable-line

import type { Dispatch } from "react";
import type { MarkdownContextReducerAction } from "types/markdown-context";

import { MarkdownContextReducerActionType } from "types/markdown-context";

export default class MarkdownParserWorker {
  private readonly worker: Worker;

  constructor(dispatch: Dispatch<MarkdownContextReducerAction>) {
    this.worker = new Worker();
    this.worker.onmessage = ({ data }: { data: HtmlArray }) =>
      dispatch({
        type: MarkdownContextReducerActionType.SetHtmlArray,
        htmlArray: data,
      });
  }

  parse(md: MarkdownString): void {
    this.worker.postMessage(md);
  }

  terminate(): void {
    if (this.worker) this.worker.terminate();
  }
}
