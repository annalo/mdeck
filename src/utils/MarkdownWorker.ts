import Worker from "worker-loader!./Worker"; // eslint-disable-line
import type { Dispatch } from "react";
import type { MarkdownContextReducerAction } from "types/markdown-context";

export class MarkdownWorker {
  private readonly worker: Worker;

  constructor(dispatch: Dispatch<MarkdownContextReducerAction>) {
    this.worker = new Worker();
    this.worker.onmessage = ({
      data,
    }: {
      data: MarkdownContextReducerAction;
    }) => {
      console.log("markdown worker on message", data);
      dispatch(data);
    };
  }

  parse(md: string): void {
    this.worker.postMessage({ action: "PARSE", md });
  }

  terminate() {
    if (this.worker) this.worker.terminate();
  }
}
