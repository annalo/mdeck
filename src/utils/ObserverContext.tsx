import { createContext } from "react";

export const ObserverContext = createContext<IntersectionObserver | null>(null);
