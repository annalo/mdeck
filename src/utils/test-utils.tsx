/* eslint-disable import/no-extraneous-dependencies */
import React from "react";
import { render } from "@testing-library/react";
import { renderHook } from "@testing-library/react-hooks";

import { ThemeProvider } from "components/App/ThemeProvider";

import { MarkdownContextProvider } from "contexts/MarkdownContext";
import { PresentationContextProvider } from "contexts/PresentationContext";
import { SlideObserverProvider } from "contexts/SlideObserver";
import { CodeLineObserverProvider } from "contexts/CodeLineObserver";

const AllTheProviders = ({ children }) => {
  return (
    <ThemeProvider>
      <MarkdownContextProvider>
        <PresentationContextProvider>
          <SlideObserverProvider>
            <CodeLineObserverProvider>{children}</CodeLineObserverProvider>
          </SlideObserverProvider>
        </PresentationContextProvider>
      </MarkdownContextProvider>
    </ThemeProvider>
  );
};

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

const customRenderHook = (hook, options) =>
  renderHook(hook, { wrapper: AllTheProviders, ...options });

// override render and renderHook methods
export { customRender as render, customRenderHook as renderHook };
