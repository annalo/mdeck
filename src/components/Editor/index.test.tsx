import React from "react";
import { fireEvent, screen } from "@testing-library/react";
import { act } from "@testing-library/react-hooks";
import userEvent from "@testing-library/user-event";
import { render, renderHook } from "utils/test-utils";

import {
  MARKDOWN_CONTEXT_DEFAULT_INITIAL_STATE,
  useMarkdownDispatch,
  useMarkdownState,
} from "contexts/MarkdownContext";
import { Editor } from ".";
import { TEXT_AREA_LINE_HEIGHT } from "./TextArea";
import { useSyncTextArea } from "./useSyncTextArea";
import { useTrackTextAreaScroll } from "./useTrackTextAreaScroll";

describe("<Editor />", () => {
  test("should render and match the snapshot", () => {
    const { asFragment } = render(<Editor />);
    expect(asFragment()).toMatchSnapshot();
  });

  test("should set md on textarea change", () => {
    render(<Editor />);
    const textarea = screen.getByRole("textbox");

    expect(textarea).toHaveValue("");
    userEvent.type(textarea, "# Hello World!");
    expect(textarea).toHaveValue("# Hello World!");
  });

  test("should support tabs in textarea", () => {
    render(<Editor />);
    const textarea = screen.getByRole("textbox");

    expect(textarea).toHaveValue("");

    userEvent.type(textarea, "Hello");
    fireEvent.keyDown(textarea, { key: "Tab", keyCode: 9 });
    userEvent.type(textarea, "World!");

    expect(textarea).toHaveValue("Hello\tWorld!");
  });

  describe("useSyncTextArea", () => {
    test("should sync textarea to slideshowLineNumber corresponding text", () => {
      const textarea = document.createElement("textarea");
      const ref = { current: textarea };
      const slideshowLineNumber = 7;

      renderHook(() => {
        useSyncTextArea({
          ref,
          slideshowLineNumber,
          textAreaLineHeight: TEXT_AREA_LINE_HEIGHT,
        });
      });

      expect(textarea.scrollTop).toBe(
        TEXT_AREA_LINE_HEIGHT * slideshowLineNumber
      );
    });
  });

  describe("useTrackTextAreaScroll", () => {
    let textarea;
    let lineCount;
    let scrollHeight;

    beforeEach(() => {
      // Fill textarea with lines of text to scroll
      textarea = document.createElement("textarea");
      lineCount = 7;
      scrollHeight = lineCount * TEXT_AREA_LINE_HEIGHT;
      const textValue = Array(...Array(lineCount))
        .map(() => "test")
        .join("\n");
      textarea.value = textValue;
    });

    test("should set textLineNumber to the top most visible line number on scroll if isActive", () => {
      const ref = { current: textarea };
      const { result } = renderHook(() => {
        const dispatch = useMarkdownDispatch();
        useTrackTextAreaScroll({
          dispatch,
          isActive: true,
          ref,
          textAreaLineHeight: TEXT_AREA_LINE_HEIGHT,
        });
        return useMarkdownState();
      });

      expect(result.current.textLineNumber).toBe(
        MARKDOWN_CONTEXT_DEFAULT_INITIAL_STATE.textLineNumber
      );

      act(() => {
        fireEvent.scroll(textarea, { target: { scrollTop: scrollHeight } });
      });

      expect(result.current.textLineNumber).toBe(lineCount);
    });

    test("should not set textLineNumber on scroll if isActive is false", () => {
      const ref = { current: textarea };
      const { result } = renderHook(() => {
        const dispatch = useMarkdownDispatch();
        useTrackTextAreaScroll({
          dispatch,
          isActive: false,
          ref,
          textAreaLineHeight: TEXT_AREA_LINE_HEIGHT,
        });
        return useMarkdownState();
      });

      act(() => {
        fireEvent.scroll(textarea, { target: { scrollTop: scrollHeight } });
      });

      expect(result.current.textLineNumber).toBe(
        MARKDOWN_CONTEXT_DEFAULT_INITIAL_STATE.textLineNumber
      );
    });
  });
});
