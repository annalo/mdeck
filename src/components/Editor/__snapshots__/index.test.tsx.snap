// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Editor /> should render and match the snapshot 1`] = `
<DocumentFragment>
  .c1 {
  background-color: transparent;
  border: none;
  height: 100%;
  outline: none;
  resize: none;
  color: #2f2f2f;
  font-size: 15px;
  margin-bottom: 0.3em;
  line-height: 18px;
  padding: 0.5em;
  tab-size: 2;
  -ms-overflow-style: none;
  -webkit-scrollbar-width: none;
  -moz-scrollbar-width: none;
  -ms-scrollbar-width: none;
  scrollbar-width: none;
}

.c1::-webkit-scrollbar {
  display: none;
}

.c0 {
  background-color: #fafafa;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<div
    class="c0"
  >
    <textarea
      class="c1"
      name="md"
    />
  </div>
</DocumentFragment>
`;
